src/blueprint.zig

```diff
@@ -0,0 +1,281 @@
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: usize = @abs(endX - startX + 1);
+		const sizeY: usize = @abs(endY - startY + 1);
+		const sizeZ: usize = @abs(endZ - startZ + 1);
+
+		self.sizeX = sizeX;
+		self.sizeY = sizeY;
+		self.sizeZ = sizeZ;
+
+		self.blocks.ensureCapacity(self.sizeX*self.sizeY*self.sizeZ);
+
+		for(0..sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const maybeBlock = main.server.world.?.getBlock(worldX, worldY, worldZ);
+					if(maybeBlock) |block| {
+						self.blocks.appendAssumeCapacity(block);
+					} else {
+						return .{.x = worldX, .y = worldY, .z = worldZ, .message = "Chunk containing block not loaded."};
+					}
+				}
+			}
+		}
+		return null;
+	}
+	pub fn paste(self: @This(), pos: Vec3i) void {
+		const startX = pos[0];
+		const startY = pos[1];
+		const startZ = pos[2];
+
+		var blockIndex: usize = 0;
+
+		for(0..self.sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..self.sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..self.sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = self.blocks.items[blockIndex];
+					_ = main.server.world.?.updateBlock(worldX, worldY, worldZ, block);
+
+					blockIndex += 1;
+				}
+			}
+		}
+	}
+	pub fn store(self: @This(), allocator: NeverFailingAllocator) []u8 {
+		std.debug.assert(self.sizeX != 0);
+		std.debug.assert(self.sizeY != 0);
+		std.debug.assert(self.sizeZ != 0);
+
+		var gameIdToBlueprintId = self.makeGameIdToBlueprintIdMap(main.stackAllocator);
+		defer gameIdToBlueprintId.deinit();
+		std.debug.assert(gameIdToBlueprintId.count() != 0);
+
+		const blockPalette = self.packBlockPalette(main.stackAllocator, gameIdToBlueprintId);
+		defer main.stackAllocator.free(blockPalette);
+		std.debug.assert(gameIdToBlueprintId.count() == blockPalette.len);
+
+		const blockPaletteSizeBytes = self.getBlockPaletteSize(blockPalette);
+		const blockArraySizeBytes = self.sizeX*self.sizeY*self.sizeZ*@sizeOf(BlockStorageType);
+		const decompressedDataSizeBytes = blockPaletteSizeBytes + blockArraySizeBytes;
+		std.debug.assert(decompressedDataSizeBytes != 0);
+
+		var decompressedWriter = BinaryWriter.initCapacity(allocator, .big, decompressedDataSizeBytes);
+		defer decompressedWriter.deinit();
+
+		self.storeBlockPalette(&decompressedWriter, blockPalette);
+		self.storeBlockArray(&decompressedWriter, gameIdToBlueprintId);
+
+		const compression: BlueprintCompression = .deflate;
+		const compressedData: []u8 = self.compressOutputBuffer(main.stackAllocator, decompressedWriter.data.items, compression);
+		defer main.stackAllocator.free(compressedData);
+		std.debug.assert(compressedData.len != 0);
```

Comment:
0 bytes of data can be a valid result of a compression algorithm. The decompression function could then e.g. just fill the entire output buffer with 0 (air).